Added expressjs-style routing

kurtmilam opened the pull request:
Most changes are in router.js.
I npm installed path-to-regexp, used by the expressjs router.
I updated routes.js to work with the new router and added a route with
a path param.
I updated header.js and main.js to show an example of a route with a
path param working (/main/:id -> /main/1234)

kurtmilam committed:
Added expressjs-style routing

Most changes are in router.js.
I npm installed path-to-regexp, used by the expressjs router.
I updated routes.js to work with the new router and added a route with
a path param.
I updated header.js and main.js to show an example of a route with a
path param working (/main/:id -> /main/1234)

polytypic commented:
Thanks, this looks very nice!

I have a couple of suggestions.  I'll attach comments to the code.

polytypic commented on src/public/pages/main.js:
```diff
@@ -26,8 +26,12 @@ const Contacts = ({contacts}) =>
-  </div>)
+    { typeof L.get( [ 'params', 'id' ], props ) !== 'undefined'
+      && <div>With ID: { L.get( [ 'params', 'id' ], props ) }</div>
+    }
```
Karet Util has functions `ift` (aka "if-then") and `ifte` (aka "if-then-else") for conditionals like this.  You could write:

```js
{ U.ift(L.get( [ 'params', 'id' ], props ),
        <div>With ID: { L.get( [ 'params', 'id' ], props ) }
```

Note that with `undefined` there is no need to perform a `typeof` test.

I would also suggest dropping the separate `params` object and just matching the `id` property directly which would simplify this whole thing to:

```js
export default U.withContext(({id}, {state}) =>
  <div>
    <h1>Main page</h1>
    <Contacts contacts={U.view(["contacts", L.define([])], state)}/>
    { U.ift(id, <div>With ID: { id }</div> }
  </div>
)
```

kurtmilam replied:
I agree with all of your suggestions. You touched on a couple of things I was uncertain about, so the clarification is helpful. I'll make, test and commit the changes shortly.

polytypic commented on src/public/components/router.js:
```diff
@@ -1,6 +1,83 @@
+// end route resolution
+
+const Router = U.withContext( resolveRoute )
+Router.init = init
```
Hmm... It doesn't really make a big difference, but I would generally recommend using named exports.  Perhaps rename `init` to e.g. `prepareRoutes`?

kurtmilam committed:
Updated expressjs-style routing based on comments by @polytopic

1. Switched to named exports in router.js (route initialization moved
from Router.init to prepareRoutes named export)
2. Send naked params as props to resolved component (rather than an
object containing several properties, among them 'params').
3. Clean up the conditional ID display on main.js using U.ift() & {id}.
4. ALSO: Switched out the ternary operator in resolveDynamic with
R.ifElse, removing the need to create a const to hold regex matches.

polytypic commented on src/public/components/router.js:
```diff
@@ -1,6 +1,83 @@
+                       { const match = regex.exec( path )
+                         return match !== null
+                                ? R.reduced( { component, path, route, regex
+                                             , params: prepParams( regex.keys )( R.tail( match ) )
```
Here I would suggest avoiding collecting the params into a `params` property and instead passing the parameters as direct, unnested, props to the component.

Note that `path` is available from the context.  I probably wouldn't pass it as a prop to the component at all.

Hmm... Do you have a specific use case in mind for passing the `route` and `regex` props to the component?  I'm not sure that the `route` and `regex` are useful props for the component.  I would probably just drop those.

So, perhaps something like this:

```js
const resolveDynamic = path =>
  R.compose( R.reduce( ( acc, { component, regex } ) =>
                        { const match = regex.exec( path )
                          return match !== null
                                 ? R.reduced( { component
                                              , props: prepParams( regex.keys )( R.tail( match ) } )
// ...
```

kurtmilam replied:
I had actually already made this change on my side, last night (although I hadn't renamed `params` to `props`. I vacillated between passing everything into the component and just passing the props/params, but figured we'd settle on just passing the props and dropping the rest.

I'll rename `params` to `props`.

polytypic commented on src/public/components/router.js:
```diff
@@ -1,6 +1,83 @@
+                      resolveStatic( path )( routes )
+                      || resolveDynamic( path )( routes )
+                      || { component: NotFound, path }
+                    return React.createElement( component, route )
```
With the above suggestion, this would become something like:

```js
const { component, props } =
  resolveStatic( path )( routes )
  || resolveDynamic( path )( routes )
  || { component: NotFound }
return React.createElement( component, props )
```

kurtmilam replied:
Agreed. I'll rename `params` to `props`. Otherwise, this is already how things look as of my previous commit :)

kurtmilam committed:
Renamed 'params' property to 'props' in resolveRoute and resolveDynamic

Minor updates to routing based on review comments by @polytopic.

polytypic commented on src/public/components/router.js:
```diff
@@ -1,6 +1,82 @@
+                  { const { component, props } =
+                      resolveStatic( path )( routes )
+                      || resolveDynamic( path )( routes )
+                      || { component: NotFound, path }
```
`path` here is now unused.

kurtmilam committed:
Removed unused 'path' property from NotFound return

In case of an unmatched route, I switched from returning
{ component: NotFound, path } to { component: NotFound } at @polytypics'
suggestion, since we're no longer passing anything but the unnested
props to the resolved component.

polytypic commented:
Ready to merge?

kurtmilam commented:
If it's alright, I'd like a few more minutes to try out a small change that'll lead to slightly cleaner code.

kurtmilam committed:
Used destructuring rather than L.get in resolveStatic, resolveDynamic

Had been using R.compose in resolveStatic and resolveDynamic purely to
allow extraction of the relevant property from the routes object
(.static and .dynamic, respectively) via L.get .

kurtmilam commented:
See what you think. I just switched to using destructuring in resolveStatic and resolveDynamic to allow me to get rid of `R.compose()` wrappers in each that were just there to let me use `L.get` to extract the relevant properties from the `routes` argument.

Was able to remove a few lines of code and a layer of indentation.

polytypic commented:
> switched to using destructuring

It looks better to me. 👍

I would personally take that approach even further.  IMHO, point free programming is a mixed bag.  Sometimes it makes things nicer.  Especially in cases where you have a simple linear flow in the code (IOW, you can just `pipe` (or `compose` - I prefer `pipe` as I like to read code [left-to-right and top-to-bottom](http://theburningmonk.com/2014/12/being-visually-honest-with-f/)) without special combinators to e.g. reuse parameters) or when you can otherwise completely ignore extra parameters.  Other times point-free style just makes things more difficult.  I often shift code to point-free style and back to see which looks better.  I would probably generally recommend avoiding point-free style unless it looks clearly better than just having named parameters.

kurtmilam commented:
Great! I can point things up a little more, if you'd like. I agree that point-free style comes with costs, and am still figuring out when, where and to what extent it makes sense to pursue a point-free style.

kurtmilam commented:
I can also switch from `compose` to `pipe`, get rid of my W-combinator and make any other style-related changes you'd prefer to see.

kurtmilam committed:
Pointy, piped and sans combinator

kurtmilam committed:
Moved a comment that was out of place after previous changes

kurtmilam commented on src/public/components/router.js:
```diff
@@ -1,6 +1,75 @@
+
+const resolveRoute = ( { routes, NotFound }, { path }  ) =>
+  U.fromKefir( K( path, path =>
+                  { const { component, props } =
```
Is there a way to get rid of this temporary `const` (at line 63) while maintaining readability or, hopefully, improving readability? As you can see, I try to avoid creating temporary container variables and using `return` statements in my arrow functions where I can.

polytypic replied:
Perhaps return an array `[Component, props]` and then pass that to `R.apply(React.createElement)`.

polytypic commented:
Refactoring the router for a moment, I ended up with this:

```js
import * as L  from "partial.lenses"
import * as R  from "ramda"
import * as U  from "karet.util"
import React   from "karet"
import toRegex from "path-to-regexp"

const addRegexes = R.map(entry => {
  const regex = toRegex(entry.route)
  regex.keys = R.map(L.get("name"), regex.keys)
  return {...entry, regex}
})

const expandRoutes =
  R.pipe(R.toPairs,
         R.map(R.zipObj(["route", "Component"])))

export const prepareRoutes = R.pipe(expandRoutes, addRegexes)

const router = U.lift(({routes, NotFound}, {path}) => {
  for (let i=0; i<routes.length; ++i) {
    const {Component, regex} = routes[i]
    const match = regex.exec(path)
    if (match)
      return <Component {...R.zipObj( regex.keys, R.tail( match ) )}/>
  }
  return <NotFound/>
})

export const Router = U.withContext(R.pipe(router, U.fromKefir))
```

Main changes:
* Removed the static-dynamic distinction
   * Reduced amount of code
   * Now all routes are matched the same regarding case-sensitivity and trailing `/`
   * Now routes are matched in definition order
* Precomputed the array of `name`s of the Regex `keys`

I just also used a simple loop.  I wish Ramda would have a function like [`tryPick`](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/list.trypick%5b't,'u%5d-function-%5bfsharp%5d) or [`L.selectAs`](https://github.com/calmm-js/partial.lenses/#L-selectAs), but without optics, because that is what I almost always want (to find an element and immediately compute something from the found element).

kurtmilam commented:
Looks good! The reasons I split the dynamic and static routes:
1. Always prefer the static route in case a dynamic and static route match the path, irrespective of the order in which the routes were defined in the raw `routes` object.
2. Potentially improve performance (or make it easy to improve performance in the future) by avoiding `regex.exec` calls on static paths.

I didn't split dynamic and static routes in my original router and ran into some cases where a dynamic and static route matched a path, and the order in which the routes had been defined determined the route that was resolved.

polytypic commented:
Note that my refactored version is just a suggestion. :)

Matching a path against the RegExp of each route in turn is roughly `O(r * p)` in the number of routes `r` and length `p` of the path and RegExp matching is probably a lot slower than string comparisons.

It would be possible to optimize routing to essentially `O(1 * p)`, which means that the number of routes wouldn't matter.  That is basically what e.g. RegExp based lexical analyzer generators do.  They compute the union of the regular expressions and match all of them simultaneously.  But without a nice library for that kind of matching it would be a bit overkill. :)

I think that functionality wise the router looks very nice and it should be good enough for many applications.  I consider this boilerplate project as an example and expect people would take it and then rewrite it to their needs.

Thanks again for the PR!  Just let me know when you feel it is ready and I'll merge it.

kurtmilam commented:
I had thought about eventually changing the static routes back to `{ '/route': component }` and switching the static Resolver to a simple `staticRoutes[ '/path' ]`, which I thought might be quicker than either a `regex.exec` or the current `R.find` that I'm using, but I agree that getting case-insensitivity and ignoring trailing slashes is probably worth more than any performance gains one might get from avoiding the `regex.exec` calls.

I guess the final question is whether you think it's worth splitting the static and dynamic routes in order to make it possible to prefer static over dynamic matches, irrespective of the order in which the routes were defined.

I also see that you're not as averse to creating temporary variables as I am :)

I'l take some time to think about your refactoring and figure out which style I find more readable (just as an exercise for myself - I have no preference over how the actual router code that's merged looks, in the end). It's entirely possible that I'm overusing ramda / fp in situations where code with a slightly more imperative favor would offer advantages in terms of concision, readability and/or performance.

Lastly, regarding computing the union of all regular expressions, I read up on that technique on Monday and considered using it in my solution. Then I decided it was probably best to get a router working and move on to improving my understanding of calmm :)

kurtmilam committed:
Switched to regex.exec for static routes

kurtmilam commented:
The previous commit switches to using the same `regex-exec`-based `resolver` function for static and dynamic routes, while still preferring static matches to dynamic ones.

polytypic commented on src/public/components/router.js:
```diff
@@ -1,6 +1,70 @@
+const expandRoutes = rawRoutes =>
+  R.pipe( R.toPairs
+        , R.map( R.zipObj( [ 'route', 'component' ] ) )
+        )( rawRoutes )
```
In cases like this I would drop the argument:

```js
const expandRoutes =
   R.pipe( R.toPairs
         , R.map( R.zipObj( [ 'route', 'component' ] ) )
         )
```

This is like the ideal case for point-free programming where one can just use simple composition.

kurtmilam replied:
👍 Same for `addRegexes`, `partitionDynamicAndStatic` and `prepareRoutes`?

polytypic commented on src/public/components/router.js:
```diff
@@ -1,6 +1,70 @@
+  R.partition( R.pipe( L.get( [ 'regex', 'keys' ] )
+                     , R.length
+                     , R.lt( 0 )
+                     )
```
IMHO, asymmetric comparisons start to get a little difficult to read in point-free style.  Compare to:

```js
({regex: {keys}}) => keys.length > 0
```

kurtmilam replied:
👍 Agreed, yours is certainly easier to read. I'm beginning to think that I may have drunk too much of the ramda kool-aid.

polytypic commented:
Note that I definitely generally prefer using referentially transparent functions.  The use of a `for` loop is just a practical choice, because there isn't a really better alternative (that I can see) in JS (most alternatives are longer and slower). In F#, I would have likely used `tryPick`.  Also, use of `for` loop inside a function is not visible outside of the function in any way.  Referential transparency is a so called extensional property of functions.

kurtmilam commented:
👍 This (not caring as much about keeping the internals point-free / using tools like `for` loops inside functions) is something I've been doing more of in my code lately, although it's certainly not evident in my initial pull request.

As long as what happens in the function stays in the function, it's OK (and sometimes/often preferable) to take some imperative shortcuts.

I'll modify my previous commit to switch to using a `for` loop and clean things up a little more based on your refactoring, maintaining the dynamic/static partitioning. I'm doing this as an exercise for myself. If you'd prefer to jettison the partitioning, I'm A-OK if you decide to go with your refactored version of the router.

Also, thanks for taking the time to review my pull request in such detail. As a self-taught programmer who's working in isolation at the moment, the feedback from an experienced software engineer is invaluable.

kurtmilam commented:
On a side note, I wonder whether the ramda team would consider a pull request that adds a `tryPick` function to their toolbox.

polytypic commented:
No problem.  I hope my comments helped.

In a previous project we had roughly this (the arrow comments added just now):

```js
// `findAs` is roughly a convenient combination of `find` and `map`.
// `findAs(x2yMaybe, xs)` returns the first truthy value returned by the
// `x2yMaybe` function when applied to elements of the `xs` array.
export const findAs = R.curry((x2yMaybe, xs) => {
  for (let i=0; i<xs.length; ++i) {
    let yMaybe = x2yMaybe(xs[i])
    if (yMaybe) // <- Might be better to compare against `undefined` rather than look for a truthy value...
      return yMaybe
  }
  return undefined // <- ...because `undefined` is indeed the default.
})
```

I'd love to have something like that in Ramda.

kurtmilam committed:
Refactored based on comments by @polytopic

I incorporated most of @polytopic's suggested changes.
I also switched from partitioning static and dynamic routes to just
sorting by them, putting all static routes in front of all dynamic
routes, thereby maintaining the preference for static routes in cases
where a path matches a static and dynamic route.

kurtmilam commented:
Very helpful, indeed -- I look forward to my next opportunity to make a pull request to one of your projects!

The previous commit is very close to your refactoring suggestion while maintaining a preference for static route matches. Rather than partitioning, I achieved static route match preference by sorting the routes array, putting static routes at the head of the list.

polytypic commented:
Thanks, it looks very nice!  I'll merge this shortly.  It is very nice to have an example of routing with parameters extracted from the path.
